File: safepath.go
// Safe path handling: zip-slip defence, Windows device names, long paths.

package main

import (
	"errors"
	"os"
	"path/filepath"
	"strings"
)

// Reserved Windows device names. A file actually named CON or NUL is nearly
// impossible to delete from Explorer, so these are always renamed.
var reservedNames = map[string]bool{
	"CON": true, "PRN": true, "AUX": true, "NUL": true,
	"COM1": true, "COM2": true, "COM3": true, "COM4": true, "COM5": true,
	"COM6": true, "COM7": true, "COM8": true, "COM9": true, "COM0": true,
	"LPT1": true, "LPT2": true, "LPT3": true, "LPT4": true, "LPT5": true,
	"LPT6": true, "LPT7": true, "LPT8": true, "LPT9": true, "LPT0": true,
	"CONIN$": true, "CONOUT$": true, "CLOCK$": true,
	"COM¹": true, "COM²": true, "COM³": true,
}

// Why an archive entry was rejected.
var (
	errTraversal = errors.New("path escapes target folder")
	errEmptyName = errors.New("empty name")
)

// cleanComponent sanitises a single path component for Windows.
func cleanComponent(part string) string {
	var b strings.Builder
	for _, c := range part {
		switch {
		case c == '<' || c == '>' || c == ':' || c == '"' || c == '|' || c == '?' || c == '*':
			b.WriteByte('_')
		case c < 32:
			b.WriteByte('_')
		default:
			b.WriteRune(c)
		}
	}
	cleaned := b.String()

	// Windows silently strips trailing spaces and dots; append '_' to keep the
	// name distinct instead of letting two entries collide.
	trimmed := strings.TrimRight(cleaned, " .")
	if len(trimmed) != len(cleaned) {
		if trimmed == "" {
			cleaned = ""
		} else {
			cleaned = trimmed + "_"
		}
	}

	// Device names are reserved even with an extension, e.g. CON.txt
	stem, _, _ := strings.Cut(cleaned, ".")
	if stem != "" && reservedNames[strings.ToUpper(stem)] {
		cleaned = "_" + cleaned
	}
	return cleaned
}

// sanitize turns an archive entry name into a safe relative path, blocking `..`
// traversal, absolute paths, drive letters, forbidden characters and reserved
// device names.
func sanitize(name string) (string, error) {
	var parts []string
	for _, raw := range strings.FieldsFunc(name, func(r rune) bool {
		return r == '/' || r == '\\'
	}) {
		if raw == "." {
			continue
		}
		if raw == ".." {
			return "", errTraversal
		}
		// Drop a leading drive specifier such as "C:"
		if len(raw) == 2 && raw[1] == ':' &&
			((raw[0] >= 'a' && raw[0] <= 'z') || (raw[0] >= 'A' && raw[0] <= 'Z')) {
			continue
		}
		cleaned := cleanComponent(raw)
		if cleaned == "" {
			continue
		}
		parts = append(parts, cleaned)
	}
	if len(parts) == 0 {
		return "", errEmptyName
	}
	return filepath.Join(parts...), nil
}

// prepareRoot creates the destination folder and returns it as an absolute
// path. Go's os package converts long absolute paths to the `\\?\` form itself,
// so paths beyond the 260-character limit work without extra handling here.
func prepareRoot(dir string) (string, error) {
	if err := os.MkdirAll(dir, 0o777); err != nil {
		return "", err
	}
	abs, err := filepath.Abs(dir)
	if err != nil {
		return dir, nil
	}
	return abs, nil
}

// displayPath strips the `\\?\` prefix when showing a path to a person.
func displayPath(p string) string {
	if rest, ok := strings.CutPrefix(p, `\\?\UNC\`); ok {
		return `\\` + rest
	}
	if rest, ok := strings.CutPrefix(p, `\\?\`); ok {
		return rest
	}
	return p
}

File: src/safepath.rs
//! Safe path handling: zip-slip defence, Windows device names, long paths.

use std::path::{Path, PathBuf};

/// Reserved Windows device names. A file actually named `CON` or `NUL` is
/// nearly impossible to delete from Explorer, so these are always renamed.
const RESERVED: [&str; 30] = [
    "CON", "PRN", "AUX", "NUL", "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8",
    "COM9", "COM0", "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9",
    "LPT0", "CONIN$", "CONOUT$", "COM\u{00b9}", "COM\u{00b2}", "COM\u{00b3}", "CLOCK$",
];

/// Why an archive entry was rejected.
#[derive(Debug, PartialEq)]
pub enum Reject {
    /// Contains `..` — a zip-slip attempt.
    Traversal,
    /// Nothing usable left after cleaning.
    Empty,
}

fn is_reserved(stem: &str) -> bool {
    let upper = stem.to_ascii_uppercase();
    RESERVED.iter().any(|r| *r == upper)
}

/// Clean a single path component for Windows.
fn clean_component(part: &str) -> String {
    let mut cleaned: String = part
        .chars()
        .map(|c| match c {
            '<' | '>' | ':' | '"' | '|' | '?' | '*' => '_',
            c if (c as u32) < 32 => '_',
            c => c,
        })
        .collect();

    // Windows silently strips trailing spaces and dots; append '_' to keep the
    // name distinct instead of letting two entries collide.
    let trimmed = cleaned.trim_end_matches([' ', '.']);
    if trimmed.len() != cleaned.len() {
        cleaned = if trimmed.is_empty() {
            String::new()
        } else {
            format!("{}_", trimmed)
        };
    }

    // Device names are reserved even with an extension, e.g. CON.txt
    let stem = cleaned.split('.').next().unwrap_or("");
    if !stem.is_empty() && is_reserved(stem) {
        cleaned = format!("_{}", cleaned);
    }
    cleaned
}

/// Turn an archive entry name into a safe relative path.
///
/// Blocks `..` traversal, absolute paths, drive letters, forbidden characters
/// and reserved device names.
pub fn sanitize(name: &str) -> Result<PathBuf, Reject> {
    let mut out = PathBuf::new();
    let mut any = false;

    for raw in name.split(['/', '\\']) {
        if raw.is_empty() || raw == "." {
            continue;
        }
        if raw == ".." {
            return Err(Reject::Traversal);
        }
        // Drop a leading drive specifier such as "C:"
        if raw.len() == 2 && raw.ends_with(':') && raw.as_bytes()[0].is_ascii_alphabetic() {
            continue;
        }
        let cleaned = clean_component(raw);
        if cleaned.is_empty() {
            continue;
        }
        out.push(cleaned);
        any = true;
    }

    if any { Ok(out) } else { Err(Reject::Empty) }
}

/// Create the destination folder and return it in extended form (`\\?\C:\...`)
/// so paths longer than the 260-character MAX_PATH limit still work.
pub fn prepare_root(dir: &Path) -> std::io::Result<PathBuf> {
    std::fs::create_dir_all(dir)?;
    // On Windows canonicalize() returns the `\\?\` form, which is what we want.
    match std::fs::canonicalize(dir) {
        Ok(c) => Ok(c),
        Err(_) => Ok(dir.to_path_buf()),
    }
}

/// Strip the `\\?\` prefix when showing a path to the user.
pub fn display(p: &Path) -> String {
    let s = p.to_string_lossy();
    if let Some(rest) = s.strip_prefix(r"\\?\UNC\") {
        format!(r"\\{}", rest)
    } else if let Some(rest) = s.strip_prefix(r"\\?\") {
        rest.to_string()
    } else {
        s.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn blocks_zip_slip() {
        assert_eq!(sanitize("../evil.txt"), Err(Reject::Traversal));
        assert_eq!(sanitize("a/../../b"), Err(Reject::Traversal));
        assert_eq!(sanitize("..\\evil"), Err(Reject::Traversal));
    }

    #[test]
    fn strips_absolute_paths() {
        assert_eq!(sanitize("C:/Windows/x.txt").unwrap(), PathBuf::from("Windows/x.txt"));
        assert_eq!(sanitize("/etc/passwd").unwrap(), PathBuf::from("etc/passwd"));
    }

    #[test]
    fn renames_device_names() {
        assert_eq!(sanitize("CON").unwrap(), PathBuf::from("_CON"));
        assert_eq!(sanitize("con.txt").unwrap(), PathBuf::from("_con.txt"));
        assert_eq!(sanitize("LPT1.dat").unwrap(), PathBuf::from("_LPT1.dat"));
        assert_eq!(sanitize("NUL").unwrap(), PathBuf::from("_NUL"));
        // Ordinary names that merely start the same are left alone
        assert_eq!(sanitize("console.log").unwrap(), PathBuf::from("console.log"));
    }

    #[test]
    fn cleans_forbidden_characters() {
        assert_eq!(sanitize("a<b>c.txt").unwrap(), PathBuf::from("a_b_c.txt"));
        assert_eq!(sanitize("name.").unwrap(), PathBuf::from("name_"));
        assert_eq!(sanitize("   ").unwrap_err(), Reject::Empty);
    }

    #[test]
    fn keeps_unicode() {
        assert_eq!(
            sanitize("\u{6587}\u{6863}/caf\u{e9}.txt").unwrap(),
            PathBuf::from("\u{6587}\u{6863}/caf\u{e9}.txt")
        );
    }
}
